man/man5/transport.5
.TH TRANSPORT 5 
.ad
.fi
.SH NAME
transport
\-
Postfix transport table format
.SH "SYNOPSIS"
.na
.nf
\fBpostmap /etc/postfix/transport\fR

\fBpostmap \-q "\fIstring\fB" /etc/postfix/transport\fR

\fBpostmap \-q \- /etc/postfix/transport <\fIinputfile\fR
.SH DESCRIPTION
.ad
.fi
The optional \fBtransport\fR(5) table specifies a mapping from email
addresses to message delivery transports and next\-hop destinations.
Message delivery transports such as \fBlocal\fR or \fBsmtp\fR
are defined in the \fBmaster.cf\fR file, and next\-hop
destinations are typically hosts or domain names. The
table is searched by the \fBtrivial\-rewrite\fR(8) daemon.

This mapping overrides the default \fItransport\fR:\fInexthop\fR
selection that is built into Postfix:
.IP "\fBlocal_transport (default: local:$myhostname)\fR"
This is the default for final delivery to domains listed
with \fBmydestination\fR, and for [\fIipaddress\fR]
destinations that match \fB$inet_interfaces\fR or
\fB$proxy_interfaces\fR. The default \fInexthop\fR destination
is the MTA hostname.
.IP "\fBvirtual_transport (default: virtual:)\fR"
This is the default for final delivery to domains listed
with \fBvirtual_mailbox_domains\fR. The default \fInexthop\fR
destination is the recipient domain.
.IP "\fBrelay_transport (default: relay:)\fR"
This is the default for remote delivery to domains listed
with \fBrelay_domains\fR. In order of decreasing precedence,
the \fInexthop\fR destination is taken from \fBrelay_transport\fR,
\fBsender_dependent_relayhost_maps\fR, \fBrelayhost\fR, or from the
recipient domain.
.IP "\fBdefault_transport (default: smtp:)\fR"
This is the default for remote delivery to other destinations.
In order of decreasing precedence, the \fInexthop\fR
destination is taken from \fBsender_dependent_default_transport_maps,
\fBdefault_transport\fR, \fBsender_dependent_relayhost_maps\fR,
\fBrelayhost\fR, or from the recipient domain.
.PP
Normally, the \fBtransport\fR(5) table is specified as a text file
that serves as input to the \fBpostmap\fR(1) command.
The result, an indexed file in \fBdbm\fR or \fBdb\fR format, is used
for fast searching by the mail system. Execute the command
"\fBpostmap /etc/postfix/transport\fR" to rebuild an indexed
file after changing the corresponding transport table.

When the table is provided via other means such as NIS, LDAP
or SQL, the same lookups are done as for ordinary indexed files.

Alternatively, the table can be provided as a regular\-expression
map where patterns are given as regular expressions, or lookups
can be directed to a TCP\-based server. In those case, the lookups
are done in a slightly different way as described below under
"REGULAR EXPRESSION TABLES" or "TCP\-BASED TABLES".
.SH "CASE FOLDING"
.na
.nf
.ad
.fi
The search string is folded to lowercase before database
lookup. As of Postfix 2.3, the search string is not case
folded with database types such as regexp: or pcre: whose
lookup fields can match both upper and lower case.
.SH "TABLE FORMAT"
.na
.nf
.ad
.fi
The input format for the \fBpostmap\fR(1) command is as follows:
.IP "\fIpattern result\fR"
When \fIpattern\fR matches the recipient address or domain, use the
corresponding \fIresult\fR.
.IP "blank lines and comments"
Empty lines and whitespace\-only lines are ignored, as
are lines whose first non\-whitespace character is a `#'.
.IP "multi\-line text"
A logical line starts with non\-whitespace text. A line that
starts with whitespace continues a logical line.
.PP
The \fIpattern\fR specifies an email address, a domain name, or
a domain name hierarchy, as described in section "TABLE
SEARCH ORDER".

The \fIresult\fR is of the form \fItransport:nexthop\fR and
specifies how or where to deliver mail. This is described in
section "RESULT FORMAT".
.SH "TABLE SEARCH ORDER"
.na
.nf
.ad
.fi
With lookups from indexed files such as DB or DBM, or from networked
tables such as NIS, LDAP or SQL, patterns are tried in the order as
listed below:
.IP "\fIuser+extension@domain transport\fR:\fInexthop\fR"
Deliver mail for \fIuser+extension@domain\fR through
\fItransport\fR to
\fInexthop\fR.
.IP "\fIuser@domain transport\fR:\fInexthop\fR"
Deliver mail for \fIuser@domain\fR through \fItransport\fR to
\fInexthop\fR.
.IP "\fIdomain transport\fR:\fInexthop\fR"
Deliver mail for \fIdomain\fR through \fItransport\fR to
\fInexthop\fR.
.IP "\fI.domain transport\fR:\fInexthop\fR"
Deliver mail for any subdomain of \fIdomain\fR through
\fItransport\fR to \fInexthop\fR. This applies only when the
string \fBtransport_maps\fR is not listed in the
\fBparent_domain_matches_subdomains\fR configuration setting.
Otherwise, a domain name matches itself and its subdomains.
.IP "\fB*\fI transport\fR:\fInexthop\fR"
The special pattern \fB*\fR represents any address (i.e. it
functions as the wild\-card pattern, and is unique to Postfix
transport tables).
.PP
Note 1: the null recipient address is looked up as
\fB$empty_address_recipient\fR@\fB$myhostname\fR (default:
mailer\-daemon@hostname).

Note 2: \fIuser@domain\fR or \fIuser+extension@domain\fR
lookup is available in Postfix 2.0 and later.
.SH "RESULT FORMAT"
.na
.nf
.ad
.fi
The lookup result is of the form \fItransport\fB:\fInexthop\fR.
The \fItransport\fR field specifies a mail delivery transport
such as \fBsmtp\fR or \fBlocal\fR. The \fInexthop\fR field
specifies where and how to deliver mail.

The transport field specifies the name of a mail delivery transport
(the first name of a mail delivery service entry in the Postfix
\fBmaster.cf\fR file).

The nexthop field usually specifies one recipient domain
or hostname. In the case of the Postfix SMTP/LMTP client,
the nexthop field may contain a list of nexthop destinations
separated by comma or whitespace (Postfix 3.5 and later).

The syntax of a nexthop destination is transport dependent.
With SMTP, specify a service on a non\-default
port as \fIhost\fR:\fIservice\fR, and disable MX (mail exchanger)
DNS lookups with [\fIhost\fR] or [\fIhost\fR]:\fIport\fR. The [] form
is required when you specify an IP address instead of a hostname.

A null \fItransport\fR and null \fInexthop\fR field means "do
not change": use the delivery transport and nexthop information
that would be used when the entire transport table did not exist.

A non\-null \fItransport\fR field with a null \fInexthop\fR field
resets the nexthop information to the recipient domain.

A null \fItransport\fR field with non\-null \fInexthop\fR field
does not modify the transport information.
.SH "EXAMPLES"
.na
.nf
.ad
.fi
In order to deliver internal mail directly, while using a
mail relay for all other mail, specify a null entry for
internal destinations (do not change the delivery transport or
the nexthop information) and specify a wildcard for all other
destinations.

.nf
     \fB\&my.domain    :\fR
     \fB\&.my.domain   :\fR
     \fB*            smtp:outbound\-relay.my.domain\fR
.fi

In order to send mail for \fBexample.com\fR and its subdomains
via the \fBuucp\fR transport to the UUCP host named \fBexample\fR:

.nf
     \fBexample.com      uucp:example\fR
     \fB\&.example.com     uucp:example\fR
.fi

When no nexthop host name is specified, the destination domain
name is used instead. For example, the following directs mail for
\fIuser\fR@\fBexample.com\fR via the \fBslow\fR transport to a mail
exchanger for \fBexample.com\fR.  The \fBslow\fR transport could be
configured to run at most one delivery process at a time:

.nf
     \fBexample.com      slow:\fR
.fi

When no transport is specified, Postfix uses the transport that
matches the address domain class (see DESCRIPTION
above).  The following sends all mail for \fBexample.com\fR and its
subdomains to host \fBgateway.example.com\fR:

.nf
     \fBexample.com      :[gateway.example.com]\fR
     \fB\&.example.com     :[gateway.example.com]\fR
.fi

In the above example, the [] suppress MX lookups.
This prevents mail routing loops when your machine is primary MX
host for \fBexample.com\fR.

In the case of delivery via SMTP or LMTP, one may specify
\fIhost\fR:\fIservice\fR instead of just a host:

.nf
     \fBexample.com      smtp:bar.example:2025\fR
.fi

This directs mail for \fIuser\fR@\fBexample.com\fR to host \fBbar.example\fR
port \fB2025\fR. Instead of a numerical port a symbolic name may be
used. Specify [] around the hostname if MX lookups must be disabled.

Deliveries via SMTP or LMTP support multiple destinations
(Postfix >= 3.5):

.nf
     \fBexample.com      smtp:bar.example, foo.example\fR
.fi

This tries to deliver to \fBbar.example\fR before trying
to deliver to \fBfoo.example\fR.

The error mailer can be used to bounce mail:

.nf
     \fB\&.example.com     error:mail for *.example.com is not deliverable\fR
.fi

This causes all mail for \fIuser\fR@\fIanything\fB.example.com\fR
to be bounced.
.SH "REGULAR EXPRESSION TABLES"
.na
.nf
.ad
.fi
This section describes how the table lookups change when the table
is given in the form of regular expressions. For a description of
regular expression lookup table syntax, see \fBregexp_table\fR(5)
or \fBpcre_table\fR(5).

Each pattern is a regular expression that is applied to the entire
address being looked up. Thus, \fIsome.domain.hierarchy\fR is not
looked up via its parent domains,
nor is \fIuser+foo@domain\fR looked up as \fIuser@domain\fR.

Patterns are applied in the order as specified in the table, until a
pattern is found that matches the search string.

The \fBtrivial\-rewrite\fR(8) server disallows regular
expression substitution of $1 etc. in regular expression
lookup tables, because that could open a security hole
(Postfix version 2.3 and later).
.SH "TCP-BASED TABLES"
.na
.nf
.ad
.fi
This section describes how the table lookups change when lookups
are directed to a TCP\-based server. For a description of the TCP
client/server lookup protocol, see \fBtcp_table\fR(5).
This feature is not available up to and including Postfix version 2.4.

Each lookup operation uses the entire recipient address once.  Thus,
\fIsome.domain.hierarchy\fR is not looked up via its parent domains,
nor is \fIuser+foo@domain\fR looked up as \fIuser@domain\fR.

Results are the same as with indexed file lookups.
.SH "CONFIGURATION PARAMETERS"
.na
.nf
.ad
.fi
The following \fBmain.cf\fR parameters are especially relevant.
The text below provides only a parameter summary. See
\fBpostconf\fR(5) for more details including examples.
.IP "\fBempty_address_recipient (MAILER\-DAEMON)\fR"
The recipient of mail addressed to the null address.
.IP "\fBparent_domain_matches_subdomains (see 'postconf -d' output)\fR"
A list of Postfix features where the pattern "example.com" also
matches subdomains of example.com,
instead of requiring an explicit ".example.com" pattern.
.IP "\fBtransport_maps (empty)\fR"
Optional lookup tables with mappings from recipient address to
(message delivery transport, next\-hop destination).
.SH "SEE ALSO"
.na
.nf
trivial\-rewrite(8), rewrite and resolve addresses
master(5), master.cf file format
postconf(5), configuration parameters
postmap(1), Postfix lookup table manager
.SH "README FILES"
.na
.nf
.ad
.fi
Use "\fBpostconf readme_directory\fR" or
"\fBpostconf html_directory\fR" to locate this information.
.na
.nf
ADDRESS_REWRITING_README, address rewriting guide
DATABASE_README, Postfix lookup table overview
FILTER_README, external content filter
.SH "LICENSE"
.na
.nf
.ad
.fi
The Secure Mailer license must be distributed with this software.
.SH "AUTHOR(S)"
.na
.nf
Wietse Venema
IBM T.J. Watson Research
P.O. Box 704
Yorktown Heights, NY 10598, USA

Wietse Venema
Google, Inc.
111 8th Avenue
New York, NY 10011, USA


postfix/src/trivial-rewrite/transport.c
/*++
/* NAME
/*	transport 3
/* SUMMARY
/*	transport mapping
/* SYNOPSIS
/*	#include "transport.h"
/*
/*	TRANSPORT_INFO *transport_pre_init(maps_name, maps)
/*	const char *maps_name;
/*	const char *maps;
/*
/*	void	transport_post_init(info)
/*	TRANSPORT_INFO *info;
/*
/*	int	transport_lookup(info, address, rcpt_domain, channel, nexthop)
/*	TRANSPORT_INFO *info;
/*	const char *address;
/*	const char *rcpt_domain;
/*	VSTRING *channel;
/*	VSTRING *nexthop;
/*
/*	void	transport_free(info);
/*	TRANSPORT_INFO * info;
/* DESCRIPTION
/*	This module implements access to the table that maps transport
/*	user@domain addresses to (channel, nexthop) tuples.
/*
/*	transport_pre_init() performs initializations that should be
/*	done before the process enters the chroot jail, and
/*	before calling transport_lookup().
/*
/*	transport_post_init() can be invoked after entering the chroot
/*	jail, and must be called before calling transport_lookup().
/*
/*	transport_lookup() finds the channel and nexthop for the given
/*	domain, and returns 1 if something was found.	Otherwise, 0
/*	is returned.
/* DIAGNOSTICS
/*	info->transport_path->error is non-zero when the lookup
/*	should be tried again.
/* SEE ALSO
/*	maps(3), multi-dictionary search
/*	strip_addr(3), strip extension from address
/*	transport(5), format of transport map
/* CONFIGURATION PARAMETERS
/*	transport_maps, names of maps to be searched.
/* LICENSE
/* .ad
/* .fi
/*	The Secure Mailer license must be distributed with this software.
/* AUTHOR(S)
/*	Wietse Venema
/*	IBM T.J. Watson Research
/*	P.O. Box 704
/*	Yorktown Heights, NY 10598, USA
/*
/*	Wietse Venema
/*	Google, Inc.
/*	111 8th Avenue
/*	New York, NY 10011, USA
/*--*/

/* System library. */

#include <sys_defs.h>
#include <string.h>

/* Utility library. */

#include <msg.h>
#include <stringops.h>
#include <mymalloc.h>
#include <vstring.h>
#include <split_at.h>
#include <dict.h>
#include <events.h>

/* Global library. */

#include <strip_addr.h>
#include <mail_params.h>
#include <mail_addr_find.h>
#include <match_parent_style.h>
#include <mail_proto.h>

/* Application-specific. */

#include "transport.h"

static int transport_match_parent_style;

#define STR(x)	vstring_str(x)

static void transport_wildcard_init(TRANSPORT_INFO *);

/* transport_pre_init - pre-jail initialization */

TRANSPORT_INFO *transport_pre_init(const char *transport_maps_name,
				           const char *transport_maps)
{
    TRANSPORT_INFO *tp;

    tp = (TRANSPORT_INFO *) mymalloc(sizeof(*tp));
    tp->transport_path = maps_create(transport_maps_name, transport_maps,
				     DICT_FLAG_LOCK | DICT_FLAG_FOLD_FIX
				     | DICT_FLAG_NO_REGSUB
				     | DICT_FLAG_UTF8_REQUEST);
    tp->wildcard_channel = tp->wildcard_nexthop = 0;
    tp->wildcard_errno = 0;
    tp->expire = 0;
    return (tp);
}

/* transport_post_init - post-jail initialization */

void    transport_post_init(TRANSPORT_INFO *tp)
{
    transport_match_parent_style = match_parent_style(VAR_TRANSPORT_MAPS);
    transport_wildcard_init(tp);
}

/* transport_free - destroy transport info */

void    transport_free(TRANSPORT_INFO *tp)
{
    if (tp->transport_path)
	maps_free(tp->transport_path);
    if (tp->wildcard_channel)
	vstring_free(tp->wildcard_channel);
    if (tp->wildcard_nexthop)
	vstring_free(tp->wildcard_nexthop);
    myfree((void *) tp);
}

/* update_entry - update from transport table entry */

static void update_entry(const char *new_channel, const char *new_nexthop,
			         const char *rcpt_domain, VSTRING *channel,
			         VSTRING *nexthop)
{

    /*
     * :[nexthop] means don't change the channel, and don't change the
     * nexthop unless a non-default nexthop is specified. Thus, a right-hand
     * side of ":" is the transport table equivalent of a NOOP.
     */
    if (*new_channel == 0) {			/* :[nexthop] */
	if (*new_nexthop != 0)
	    vstring_strcpy(nexthop, new_nexthop);
    }

    /*
     * transport[:[nexthop]] means change the channel, and reset the nexthop
     * to the default unless a non-default nexthop is specified.
     */
    else {
	vstring_strcpy(channel, new_channel);
	if (*new_nexthop != 0)
	    vstring_strcpy(nexthop, new_nexthop);
	else if (strcmp(STR(channel), MAIL_SERVICE_ERROR) != 0
		 && strcmp(STR(channel), MAIL_SERVICE_RETRY) != 0)
	    vstring_strcpy(nexthop, rcpt_domain);
	else
	    vstring_strcpy(nexthop, "Address is undeliverable");
    }
}

/* parse_transport_entry - parse transport table entry */

static void parse_transport_entry(const char *value, const char *rcpt_domain,
				          VSTRING *channel, VSTRING *nexthop)
{
    char   *saved_value;
    const char *host;

#define FOUND		1
#define NOTFOUND	0

    /*
     * It would be great if we could specify a recipient address in the
     * lookup result. Unfortunately, we cannot simply run the result through
     * a parser that recognizes "transport:user@domain" because the lookup
     * result can have arbitrary content (especially in the case of the error
     * mailer).
     */
    saved_value = mystrdup(value);
    host = split_at(saved_value, ':');
    update_entry(saved_value, host ? host : "", rcpt_domain, channel, nexthop);
    myfree(saved_value);
}

/* transport_wildcard_init - (re) initialize wild-card lookup result */

static void transport_wildcard_init(TRANSPORT_INFO *tp)
{
    VSTRING *channel = vstring_alloc(10);
    VSTRING *nexthop = vstring_alloc(10);
    const char *value;

    /*
     * Both channel and nexthop may be zero-length strings. Therefore we must
     * use something else to represent "wild-card does not exist". We use
     * null VSTRING pointers, for historical reasons.
     */
    if (tp->wildcard_channel)
	vstring_free(tp->wildcard_channel);
    if (tp->wildcard_nexthop)
	vstring_free(tp->wildcard_nexthop);

    /*
     * Technically, the wildcard lookup pattern is redundant. A static map
     * (keys always match, result is fixed string) could achieve the same:
     * 
     * transport_maps = hash:/etc/postfix/transport static:xxx:yyy
     * 
     * But the user interface of such an approach would be less intuitive. We
     * tolerate the continued existence of wildcard lookup patterns because
     * of human interface considerations.
     */
#define WILDCARD	"*"
#define FULL		0
#define PARTIAL		DICT_FLAG_FIXED

    if ((value = maps_find(tp->transport_path, WILDCARD, FULL)) != 0) {
	parse_transport_entry(value, "", channel, nexthop);
	tp->wildcard_errno = 0;
	tp->wildcard_channel = channel;
	tp->wildcard_nexthop = nexthop;
	if (msg_verbose)
	    msg_info("wildcard_{chan:hop}={%s:%s}",
		     vstring_str(channel), vstring_str(nexthop));
    } else {
	tp->wildcard_errno = tp->transport_path->error;
	vstring_free(channel);
	vstring_free(nexthop);
	tp->wildcard_channel = 0;
	tp->wildcard_nexthop = 0;
    }
    tp->expire = event_time() + 30;		/* XXX make configurable */
}

/* transport_lookup - map a transport domain */

int     transport_lookup(TRANSPORT_INFO *tp, const char *addr,
			         const char *rcpt_domain,
			         VSTRING *channel, VSTRING *nexthop)
{
    char   *ratsign = 0;
    const char *value;

#define STREQ(x,y)	(strcmp((x), (y)) == 0)
#define DISCARD_EXTENSION ((char **) 0)

    /*
     * The null recipient is rewritten to the local mailer daemon address.
     */
    if (*addr == 0) {
	msg_warn("transport_lookup: null address - skipping table lookup");
	return (NOTFOUND);
    }

    /*
     * Look up the full and extension-stripped address, then match the domain
     * and subdomains. Try the external form before the backwards-compatible
     * internal form.
     */
#define LOOKUP_STRATEGY \
	(MA_FIND_FULL | MA_FIND_NOEXT | MA_FIND_DOMAIN | \
	(transport_match_parent_style == MATCH_FLAG_PARENT ? \
		MA_FIND_PDMS : MA_FIND_PDDMDS))

    if ((ratsign = strrchr(addr, '@')) == 0 || ratsign[1] == 0)
	msg_panic("transport_lookup: bad address: \"%s\"", addr);

    if ((value = mail_addr_find_strategy(tp->transport_path, addr, (char **) 0,
					 LOOKUP_STRATEGY)) != 0) {
	parse_transport_entry(value, rcpt_domain, channel, nexthop);
	return (FOUND);
    }
    if (tp->transport_path->error != 0)
	return (NOTFOUND);

    /*
     * Fall back to the wild-card entry.
     */
    if (tp->wildcard_errno || event_time() > tp->expire)
	transport_wildcard_init(tp);
    if (tp->wildcard_errno) {
	tp->transport_path->error = tp->wildcard_errno;
	return (NOTFOUND);
    } else if (tp->wildcard_channel) {
	update_entry(STR(tp->wildcard_channel), STR(tp->wildcard_nexthop),
		     rcpt_domain, channel, nexthop);
	return (FOUND);
    }

    /*
     * We really did not find it.
     */
    return (NOTFOUND);
}

#ifdef TEST

 /*
  * Proof-of-concept test program. Read an address from stdin, and spit out
  * the lookup result.
  */

#include <string.h>

#include <mail_conf.h>
#include <vstream.h>
#include <vstring_vstream.h>

static NORETURN usage(const char *progname)
{
    msg_fatal("usage: %s [-v] database", progname);
}

int     main(int argc, char **argv)
{
    VSTRING *buffer = vstring_alloc(100);
    VSTRING *channel = vstring_alloc(100);
    VSTRING *nexthop = vstring_alloc(100);
    TRANSPORT_INFO *tp;
    char   *bp;
    char   *addr_field;
    char   *rcpt_domain;
    char   *expect_channel;
    char   *expect_nexthop;
    int     status;
    int     ch;
    int     errs = 0;

    /*
     * Parse JCL.
     */
    while ((ch = GETOPT(argc, argv, "v")) > 0) {
	switch (ch) {
	case 'v':
	    msg_verbose++;
	    break;
	default:
	    usage(argv[0]);
	}
    }
    if (argc != optind + 1)
	usage(argv[0]);

    /*
     * Initialize.
     */
#define UPDATE(var, val) do { myfree(var); var = mystrdup(val); } while (0)

    mail_conf_read();				/* XXX eliminate dependency. */
    UPDATE(var_rcpt_delim, "+");
    UPDATE(var_mydomain, "localdomain");
    UPDATE(var_myorigin, "localhost.localdomain");
    UPDATE(var_mydest, "localhost.localdomain");

    tp = transport_pre_init("transport map", argv[optind]);
    transport_post_init(tp);

    while (vstring_fgets_nonl(buffer, VSTREAM_IN)) {
	bp = STR(buffer);

	/*
	 * Parse the input and expectations. XXX We can't expect empty
	 * fields, so require '-' instead.
	 */
	if ((addr_field = mystrtok(&bp, ":")) == 0)
	    msg_fatal("no address field");
	if ((rcpt_domain = strrchr(addr_field, '@')) == 0)
	    msg_fatal("no recipient domain");
	rcpt_domain += 1;
	expect_channel = mystrtok(&bp, ":");
	expect_nexthop = mystrtok(&bp, ":");
	if ((expect_channel != 0) != (expect_nexthop != 0))
	    msg_fatal("specify both channel and nexthop, or specify neither");
	if (expect_channel) {
	    if (strcmp(expect_channel, "-") == 0)
		*expect_channel = 0;
	    if (strcmp(expect_nexthop, "-") == 0)
		*expect_nexthop = 0;
	    vstring_strcpy(channel, "DEFAULT");
	    vstring_strcpy(nexthop, rcpt_domain);
	}
	if (mystrtok(&bp, ":") != 0)
	    msg_fatal("garbage after nexthop field");

	/*
	 * Lookups.
	 */
	status = transport_lookup(tp, addr_field, rcpt_domain,
				  channel, nexthop);

	/*
	 * Enforce expectations.
	 */
	if (expect_nexthop && status) {
	    vstream_printf("%s:%s -> %s:%s \n",
			   addr_field, rcpt_domain,
			   STR(channel), STR(nexthop));
	    vstream_fflush(VSTREAM_OUT);
	    if (strcmp(expect_channel, STR(channel)) != 0) {
		msg_warn("expect channel '%s' but got '%s'",
			 expect_channel, STR(channel));
		errs = 1;
	    }
	    if (strcmp(expect_nexthop, STR(nexthop)) != 0) {
		msg_warn("expect nexthop '%s' but got '%s'",
			 expect_nexthop, STR(nexthop));
		errs = 1;
	    }
	} else if (expect_nexthop && !status) {
	    vstream_printf("%s:%s -> %s\n", addr_field, rcpt_domain,
			   tp->transport_path->error ?
			   "(try again)" : "(not found)");
	    vstream_fflush(VSTREAM_OUT);
	    msg_warn("expect channel '%s' but got none", expect_channel);
	    msg_warn("expect nexthop '%s' but got none", expect_nexthop);
	    errs = 1;
	} else if (!status) {
	    vstream_printf("%s:%s -> %s\n", addr_field, rcpt_domain,
			   tp->transport_path->error ?
			   "(try again)" : "(not found)");
	}
    }
    transport_free(tp);
    vstring_free(nexthop);
    vstring_free(channel);
    vstring_free(buffer);
    exit(errs != 0);
}

#endif
